### PR TITLE
[GPU] Fix synchronization issue from wrong stream in multi-stream mode on dGPU

### PR DESCRIPTION
**Pull request by andrew-k-park**
### Details:
 - Update to store network pointer in `kernel_impl_params` and use mem_lock through stream from network at runtime for `calc_output_layouts`

### Tickets:
 - 105711

### CODE REVIEW

File: src/plugins/intel_gpu/include/intel_gpu/graph/kernel_impl_params.hpp [hunk 1]
```diff
@@ -53,6 +54,7 @@ struct kernel_impl_params {
     size_t primary_input_idx = 0;
 
     memory::ptr reordered_weights = nullptr;
+    network *net = nullptr;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/graph/kernel_impl_params.hpp:57]: network::ptr can be used.

Comment (yeonbok) [reply src/plugins/intel_gpu/include/intel_gpu/graph/kernel_impl_params.hpp:57]: Or isn't it enough to have stream::ptr?

Comment (andrew-k-park) [reply src/plugins/intel_gpu/include/intel_gpu/graph/kernel_impl_params.hpp]: Yes, like 2nd option, `stream::ptr` is just enough. updated as your suggestion.

File: src/plugins/intel_gpu/src/graph/arg_max_min.cpp [hunk 2]
```diff
@@ -96,7 +96,7 @@ std::vector<layout> arg_max_min_inst::calc_output_layouts(arg_max_min_node const
     } else if (constant_mem.count(1)) {
         std::map<size_t, ngraph::HostTensorPtr> const_data;
         auto target_shape_mem = constant_mem.at(1);
-        cldnn::mem_lock<uint8_t, mem_lock_type::read> target_shape_lock(target_shape_mem, impl_param.prog->get_stream());
+        cldnn::mem_lock<uint8_t, mem_lock_type::read> target_shape_lock(target_shape_mem, impl_param.get_stream());
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/graph/arg_max_min.cpp:99]: It looks like the same changes should be applied to `src/plugins/intel_gpu/src/graph/non_max_suppression.cpp` and
`src/plugins/intel_gpu/src/graph/impls/ocl/resample.cpp`

Comment (andrew-k-park) [reply src/plugins/intel_gpu/src/graph/arg_max_min.cpp:99]: Applied to them. thanks!

**Review by yeonbok — APPROVED** [review]

**Review by e-ddykim — APPROVED** [review]
Looks good to me.

**Pull request merged**